### ISSUE TITLE
Change to Dark Theme not updating background.

### ISSUE THREAD

**Issue by ricaun** [jeremytammik/RevitLookup#194]
I updated the latest version and tested it in Revit 2024 and noticed that when changed to theme Dark the background color is not changing.

If I close and open works fine.

Here is a video.

https://github.com/jeremytammik/RevitLookup/assets/12437519/bc55d358-6f35-4397-8a48-e82a6417b3ae

**Comment by Nice3point** [comment]
What version of OS and Revit ?

**Comment by ricaun** [comment]
Windows 10 - 10.0.19045 Build 19045
Revit 24.0.4.427 - 20230308_1635(x64)

I guess the last version I tested It's needed to close the dialog to apply the theme.

**Comment by Nice3point** [comment]
Most likely the problem is related to windows 10. Unfortunately, both pomianowski (wpfui owner) and I have windows 11. I don't know if this bug will be fixed or not, but I won't be able to reproduce it, you can try to fix it if you have time

**Comment by Nice3point** [comment]
Previously the theme and applied when opening a new window. But now I have added resource transfer to NavigationService when navigating in Frame. On Win 11 it works, on Win 10 it doesn't for some reason

**Comment by ricaun** [comment]
The `RevitLookup.UI.Demo` fails as well, gonna be easier to fix.

**Comment by Nice3point** [comment]
2.2 is very different from the current one, I use 3.0 preview. Thanks for the help in debugging

**Comment by Nice3point** [comment]
I think something needs to be done in NavigationView which inherits from Frame. All in dll applications there is such a problem that Page does not inherit resources from Window. Wpf ui was designed for exe applications where this is not observed. So as you show in the video the theme does not apply specifically to Page, Window is fine. So in this code https://github.com/jeremytammik/RevitLookup/blob/b65685f5088b730f52d560e5bf064699979c3d37/RevitLookup.UI/Controls/NavigationView/NavigationView.Navigation.cs#L300-L308 when navigating I directly assign the resource dictionary from the window to the Page. I didn't think it wouldn't work on WIn 10, because there are no WinAPI calls. Maybe IF checks are not executed, or something else

**Comment by Nice3point** [comment]
events is certainly not the most efficient approach. maybe try to debug Apply https://github.com/ricaun/RevitLookup/blob/3917b7b5f2d88a076e03b7daf854352e8e240b85/RevitLookup/ViewModels/Pages/SettingsViewModel.cs#L54 method ? it contains a lot of OS dependent code 
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/04086b86-10da-43af-9ffb-c324860164ba)

**Comment by ricaun** [comment]
I really don't know the reason in my machine the windows Ui does not update probably. Maybe is missing some trigger in the OS that force the window to update the resource.

I'm testing the preview and was able to make work inside Revit, in my implementation I'm using the Changed event to update the resource in each Window/Page. Without changing to much in the main `Wpf.Ui` that's the way to do it, I guess.

Here is my PR. https://github.com/lepoco/wpfui/pull/851

I gonna build a package and try swap the `RevitLookUp.UI` to see if works.

**Comment by Nice3point** [comment]
This is not a primary problem, as very few people are still using WIn10, and losing performance to Win11 is not desirable for this purpose

**Comment by ricaun** [comment]
> This is not a primary problem, as very few people are still using WIn10, and losing performance to Win11 is not desirable for this purpose

I tested on another machine with Windows 10 and had the same problem (tested in the Hosted Revit Preview Machine).

> I don't really like the crutch of subscribing to events. I'm sure there is a more elegant solution. For now, for Win10 we can revert to the old code that changed the theme when new windows were opened
> 
> ```cs
>     partial void OnThemeChanged(ApplicationTheme value)
>     {
>         settingsService.Theme = value;
>         if (Wpf.Ui.Win32.Utilities.IsOSWindows11OrNewer)
>         {
>             ApplicationThemeManager.Apply(settingsService.Theme, settingsService.Background);
>             return;
>         }
> 
>         notificationService.ShowSuccess("Theme changed", "Changes will take effect for new windows");
>     }
> ```

Yes, this approach is much better.

**Comment by Nice3point** [comment]
@ricaun Hi Luiz, I've moved all the changes from the wpf.ui repository, there were some windows10 related fixes there recently, if you have time can you check the latest build from the dev branch? Specifically try commenting out the lines and leave only ```ApplicationThemeManager.Apply(settingsService.Theme, settingsService.Background);``` I wonder if this helped with changing the theme in Runtime https://github.com/jeremytammik/RevitLookup/blob/5301493d00ebd01c527fe388109f3ebe1541a2f3/RevitLookup/ViewModels/Pages/SettingsViewModel.cs#L64-L70

**Comment by ricaun** [comment]
Yes, it works! The theme changes in Runtime.

https://github.com/jeremytammik/RevitLookup/assets/12437519/922b4e17-d769-4c61-aa82-b7e879938678

Now you already released, so the next version 😀

Merry Christmas 🎅🎄

**Comment by ricaun** [comment]
Or not... I was messing with the Acrylic and Blur...

https://github.com/jeremytammik/RevitLookup/assets/12437519/cf67def4-a8c6-4768-a768-3a12718035d6

**Comment by Nice3point** [comment]
Oh great, that problem solved 🙃 What if in the `OnBackgroundChangedwindow` method instead of `WindowBackdropType = value;` write `ApplicationThemeManager.Apply(settingsService.Theme, settingsService.Background);` ? Ideally this should give the same result in the end, but how does it behave on your OS? In the right way, nothing should happen when you switch. All these backgrounds are not available on win10, maybe disable them altogether 🙈

Merry Christmas 🎅

**Comment by ricaun** [comment]
In the last Release, the Acrylic breaks the Dark UI.
https://github.com/jeremytammik/RevitLookup/assets/12437519/894222da-1657-437e-acca-815b2288d297

And I found this after installing the last Release. An extra `ru` folder inside 2024, there is the same `ru` folder inside the RevitLookUp.

![image](https://github.com/jeremytammik/RevitLookup/assets/12437519/5d74647c-473d-487e-85c7-8569e8e1343d)

**Comment by ricaun** [comment]
The `Acrylic` should work in `IsOSWindows7OrNewer`, in the Light version, I don't see any change. Change to `IsOSWindows11OrNewer`, same as the `Mica`.

https://github.com/jeremytammik/RevitLookup/blob/1bea5226852349f5a9b215e8dd3df4c22f2661f8/source/RevitLookup.UI/Controls/Window/WindowBackdrop.cs#L29C1-L29C79

**Comment by Nice3point** [comment]
> And I found this after installing the last Release. An extra `ru` folder inside 2024, there is the same `ru` folder inside the RevitLookUp.

I need to check where it came from

**Comment by Nice3point** [comment]
I think it's worth disabling all background effects in Windows 10 for now as they don't work there,maybe they will be fixed in wpf.ui in time. And unfortunately I can't test this, so I'm using 11

**Comment by Nice3point** [comment]
By the end of the week I think to publish Release with fixes, but in the meantime it's worth to look for any more bugs

**Comment by ricaun** [comment]
> > And I found this after installing the last Release. An extra `ru` folder inside 2024, there is the same `ru` folder inside the RevitLookUp.
> 
> I need to check where it came from

Probably in the installation, the ru is not moved or deleted. 

I don't like to use these resources for multiple languages, the extra folder/dll is annoying. Most of the time I create a json/object and swap depending on the **CurrentUICulture**. 

The **ColorRepresentationUtils** already have a **KnownColors** _Dictionary_, and swapping should be easy.

I could create a PR if you think that makes sense, with a JSON at least gonna be easier to translate to other languages 😀

**Comment by ricaun** [comment]
> By the end of the week I think to publish Release with fixes, but in the meantime it's worth to look for any more bugs
> 
> @ricaun In general, keeps the Runtime theme switching, but turn off all background effects?

Yes, turn off all the background effects in Windows 10. 

I gonna test the UI.Demo it in the Autodesk machine to see if works fine.

**Comment by Nice3point** [comment]
> Probably in the installation, the ru is not moved or deleted.

There was a problem with the installer, I wrote code that handled subfolders badly and duplicated them to the root directory. this has now been fixed. this was not observed during debugging or compiling with Nuke

**Comment by Nice3point** [comment]
> Most of the time I create a json/object and swap depending on the **CurrentUICulture**.

I wonder what it looks like, never seen anything like it before. Just resx files are basic for localisation of WPF applications, they are supported by Vs and Rider

I can also import them from csv
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/c461444d-5a01-4f24-9616-2e4f6fcb7317)


I can add comments to mark where this resource is used, it's very handy.
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/3431b3f4-28d7-416a-8692-6c15b271abdb)

I think CSV format is preferable as it has a table representation

[Colors.csv](https://github.com/jeremytammik/RevitLookup/files/13773984/Colors.csv)

![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/f6d527e2-b73c-400f-ba12-04bbb621cba4)

How convenient and efficient is it to use JSON?

**Comment by Nice3point** [comment]
> Yes, turn off all the background effects in Windows 10.

Turned it off in the last commit, you should have it like in the screenshot
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/b3633f87-f772-48b1-9413-7632224fbfb7)

**Comment by ricaun** [comment]
> > Yes, turn off all the background effects in Windows 10.
> 
> Turned it off in the last commit, you should have it like in the screenshot.

Oh, you remove the background effects options, neat!
I thought you gonna change `WindowBackdropType.Acrylic => Win32.Utilities.IsOSWindows7OrNewer,` to `IsOSWindows11OrNewer`. 😀

**Comment by Nice3point** [comment]
and why display them if they are not supported?

**Comment by ricaun** [comment]
> I wonder what it looks like, never seen anything like it before 🙈😊. Just resx files are basic for localisation of WPF applications, they are supported by Vs and Rider
> 
> I can also import them from csv

I'm not sure if Visual Studio has this csv import, never needed something like that.

> 
> How convenient and efficient is it to use JSON?

The structure would be something like this: [KnownColors.json](https://github.com/jeremytammik/RevitLookup/files/13774245/KnownColors.json)


```json
{
  "000000": "Black",
  "000080": "Navy blue",
  "0000FF": "Blue",
  "FFFFCC": "Conditioner",
  "FFFFFF": "White"
}
```

Just deserialize to a `Dictionary<string,string>`, and is good to go! That only makes sense because the code has the `KnownColors` static Dictionary, is easy to swap a **KnownColors.json** for each language.

You could store the file **KnownColors.json** in the Resources, better than each key separately.

_The `Localization.Colors` resources in not `Encoding.UTF8` there are some strange characters._

**Comment by Nice3point** [comment]
> Just deserialize to a `Dictionary<string,string>`, and is good to go!

I would still prefer to use Resx files, it is still a native way of localisation, and especially if we need to localise Xaml files or parts of it, Json is not suitable for such tasks, and it is bad practice to have several ways of localisation 🙈

**Comment by ricaun** [comment]
The only reason to use JSON is that is convenient to convert to c# object.

Instead of using:
```C#
private static readonly Dictionary<string, string> KnownColors = new()
{
    {"000000", Resources.Localization.Colors._000000},
    {"000080", Resources.Localization.Colors._000080},
    ...
    {"FFFFFF", Resources.Localization.Colors.FFFFFF},
};
```
Would be something like:
```C#
private static readonly Dictionary<string, string> KnownColors =
    Encoding.UTF8.GetString(Resources.Localization.Colors.KnownColors)
        .ToJson<Dictionary<string, string>>();
```

Still gonna have the `Colors` resources but instead of having each color key separated, you gonna have one file.

**Comment by Nice3point** [comment]
@ricaun I'll talk to the Microsoft guys about it. In the meantime we need to see if what we discussed above is fixed. If you have time, you can check out the latest dev build, and we'll roll out the Release on Friday

**Comment by ricaun** [comment]
> @ricaun I'll talk to the Microsoft guys about it. In the meantime we need to see if what we discussed above is fixed. If you have time, you can check out the latest dev build, and we'll roll out the Release on Friday

The background effect is working.

And dark background is not updating the other windows, I don't remember if was working before... 

https://github.com/jeremytammik/RevitLookup/assets/12437519/79c09af6-1ea4-4eee-b41e-6c06315e5cb9

Maybe is better to just add the `if (Wpf.Ui.Win32.Utilities.IsOSWindows11OrNewer)` like the last version.

```C#
 partial void OnThemeChanged(ApplicationTheme value)
{
    settingsService.Theme = value;
    if (Wpf.Ui.Win32.Utilities.IsOSWindows11OrNewer)
    {
        ApplicationThemeManager.Apply(settingsService.Theme, settingsService.Background);
        return;
    }

    notificationService.ShowSuccess("Theme changed", "Changes will take effect for new windows");
}
```

**Comment by Nice3point** [comment]
previous windows will not be updated, only new windows and the current window 🙃I didn't write the logic for updating old instances

**Comment by ricaun** [comment]
I assume that was already done. 

You could use the Theme event to trigger the other windows to swap the resources. Is visually pleasing when you change the theme and all windows themes change as well. 😀

**Comment by Nice3point** [comment]
You can do it if you want to)

**Comment by ricaun** [comment]
> You can do it if you want to)

I did a PR #200.

I update the `ApplicationThemeManager.Apply` to update each `Application.Windows` in there.

**Comment by Nice3point** [comment]
Then I reckon we're ready to make the last release of this year)

**Comment by Nice3point** [comment]
Only I noticed one bug after your comit, now the windows open backwards 🙈
![Uploading изображение.png…]()

**Comment by Nice3point** [comment]
![изображение](https://github.com/jeremytammik/RevitLookup/assets/20504884/185fc233-b5ce-44f8-b222-88d26f8a1812)
You can just put this loop in the ViewModel, this ensures that the theme is applied to all windows only once and not every time a new instance of RevitLookup is opened

**Comment by Nice3point** [comment]
The result is the same, only we don't touch the UI library) which makes it easier to port patches
https://github.com/jeremytammik/RevitLookup/assets/20504884/68094469-4a72-4281-adcb-789876c3b6d7

**Comment by ricaun** [comment]
> ![изображение](https://private-user-images.githubusercontent.com/20504884/293283965-d4841d76-cefe-42cd-aa64-ffeb3a14a52d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM4MDk0NjgsIm5iZiI6MTcwMzgwOTE2OCwicGF0aCI6Ii8yMDUwNDg4NC8yOTMyODM5NjUtZDQ4NDFkNzYtY2VmZS00MmNkLWFhNjQtZmZlYjNhMTRhNTJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyMzEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI5VDAwMTkyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2MWNjY2JlZDI1ZWNhODc1ZDI3YjFkZDczNWMxNmMzNGNkODAyMjY3YzFiNTAwMDlmNzE4Y2NjNzdjMGI4MTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8M9zsB6TowegYFY2eKin6SpiwNMYavHAiaXWBwE1074)
> 
> You can just put this loop in the ViewModel, this ensures that the theme is applied to all windows only once and not every time a new instance of RevitLookup is opened

Yes, that's looks great 😃
